src/languageSupport/languages/sql/connection.ts

```diff
@@ -0,0 +1,240 @@
+      return
+    }
+    const {
+      start: {line: startLine},
+      end: {line: endLine},
+    } = this._compositionRange
+
+    const changeInBlock =
+      change.range.startLineNumber >= startLine &&
+      change.range.endLineNumber <= endLine
+
+    const isDeletion = change.text == ''
+    let deletionFromBlock = false
+    if (isDeletion) {
+      const linesDeleted =
+        change.range.endLineNumber - change.range.startLineNumber
+      deletionFromBlock =
+        change.range.startLineNumber >= startLine &&
+        change.range.endLineNumber <= endLine + linesDeleted
+    }
+
+    return changeInBlock || deletionFromBlock
+  }
+
+  _setCompositionHandlers() {
+    this._model.onDidChangeContent(e => {
+      const shouldEndSync = e.changes.some(
+        change =>
+          this._editorChangeIsWithinComposition(change) &&
+          !this._couldBeFromComposition(change)
+      )
+      if (shouldEndSync) {
+        // use setTimeout to remove race condition
+        // have changes propogation first to SqlEditorMonaco.onChange()
+        setTimeout(() => {
+          this._callbackSetSession({
+            composition: {synced: false},
+          })
+          this._dispatcher(notify(compositionEnded()))
+        }, 0)
+      }
+    })
+  }
+
+  _buildComposition(
+    _toAdd: Partial<CompositionSelection>,
+    _toRemove: Partial<CompositionSelection> = null
+  ) {
+    let composition = ['SELECT *']
+    let lines = 1
+    if (this._session.measurement) {
+      composition.push(`FROM "${this._session.measurement}"`)
+      lines++
+    }
+
+    const fieldsExpr = this._session.fields
+      .map(f => `"${f}" IS NOT NULL`)
+      .join(' OR ')
+    const tagValuesExpr = Object.entries(
+      groupedTagValues(this._session.tagValues)
+    )
+      .map(
+        ([key, values]) =>
+          `"${key}" IN (${values.map(v => castType(v)).join(',')})`
+      )
+      .join(' AND ')
+
+    const timeRangeExpr = buildTimeRange(this._timeRange)
+
+    switch (
+      `${this._session.fields.length > 0}|${this._session.tagValues.length > 0}`
+    ) {
+      case `true|true`:
+        composition = composition.concat([
+          'WHERE',
+          timeRangeExpr,
+          'AND',
+          `(${fieldsExpr})`,
+          'AND',
+          tagValuesExpr,
+        ])
+        lines += 6
+        break
+      case `true|false`:
+        composition = composition.concat([
+          'WHERE',
+          timeRangeExpr,
+          'AND',
+          `(${fieldsExpr})`,
+        ])
+        lines += 4
+        break
+      case `false|true`:
+        composition = composition.concat([
+          'WHERE',
+          timeRangeExpr,
+          'AND',
+          tagValuesExpr,
+        ])
+        lines += 4
+        break
+      default:
+        composition = composition.concat(['WHERE', timeRangeExpr])
+        lines += 2
+        break
+    }
+    return {
+      composition: composition.join('\n'),
+      lines,
```

Comment:
Instead of counting `lines`, why not just use `composition.length`?

Reply:
That's more of a style choice. As the code is written now in `_buildComposition()`, the `lines` count does match the length of the composition array (prior to join). I was just being pedantic about making sure future us/devs are aware of how the line counting works.